response.js

```diff
@@ -0,0 +1,106 @@
+  // make sure query and path params always exist:
+  req.queryStringParameters = req.queryStringParameters || {};
+  req.pathParameters = req.pathParameters || {};
+  // preserve the original headers
+  req.headersRaw = Object.assign({}, req.headers);
+  // and normalize the headers so you don't have to worry about case:
+  req.headers = Object.keys(req.headersRaw).reduce((memo, key) => {
+    memo[key.toLowerCase()] = req.headersRaw[key];
+    return memo;
+  }, {});
+  if (req.headers['content-type'] === 'application/json' && req.body) {
+    try {
+      req.bodyRaw = req.body;
+      req.body = JSON.parse(new Buffer(req.body, 'base64').toString());
+    } catch (e) {
+      // do nothing
+    }
+  }
+};
+
+// middlewares must run sequentially and abort if any returns a response:
+/* eslint-disable no-await-in-loop */
+const middlewareBefore = async(req, options) => {
+  if (!options.middleware) {
+    return;
+  }
+  for (let i = 0; i < options.middleware.length; i++) {
+    const m = options.middleware[i];
+    if (m.before) {
+      const res = await m.before(req, options);
+      if (res !== undefined) {
+        return res;
+      }
+    }
+  }
+};
+
+const middlewareAfter = async(req, res, options) => {
+  if (!options.middleware) {
+    return;
+  }
+  for (let i = 0; i < options.middleware.length; i++) {
+    const m = options.middleware[i];
+    if (m.after) {
+      const afterResponse = await m.after(req, res, options);
+      if (afterResponse !== undefined) {
+        return afterResponse;
+      }
+    }
+  }
+};
+
+const runHandler = (requestHandler, req, options) => {
+  try {
+    return requestHandler(req);
+  } catch (e) {
+    log(['error'], e);
+    return {
+      statusCode: 500,
+      body: 'Server error'
+    };
+  }
+};
+
+module.exports = function(requestHandler, options = {}) {
+  // default is true:
+  const redirectTrailingSlash = options.redirectTrailingSlash === undefined ? true : options.redirectTrailingSlash;
+  return async function(req) {
+    const method = req.httpMethod || req.method;
+    // remove any trailing slash from path if it isn't '/':
+    if (redirectTrailingSlash && method.toLowerCase() === 'get' && req.path.endsWith('/') && req.path !== '/') {
+      return reply.redirect(req.path.replace(/\/$/, ''));
+    }
+    const start = new Date().getTime();
+    normalizeHeaders(req, options);
+    // if any "before" middleware returns a response, just short-circuit and return it:
+    let res = await middlewareBefore(req, options);
+    if (res !== undefined) {
+      return res;
+    }
+    // the main request handler:
+    res = await runHandler(requestHandler, req, options);
+    // if any "after" middleware returns a response, return it instead:
+    const afterResponse = await middlewareAfter(req, res, options);
+    if (afterResponse !== undefined) {
+      return afterResponse;
+    }
+    const finish = new Date().getTime();
+    const duration = finish - start;
+    const logObject = {
```

Comment:
any way we can keep getting logging even if middleware takes over?  maybe move the middleware stuff inside runHandler?